homeassistant/components/rainmachine/sensor.py

```diff
@@ -163,6 +185,109 @@ async def async_setup_entry(
+
+    def __init__(
+        self,
+        entry: ConfigEntry,
+        coordinator: DataUpdateCoordinator,
+        controller: Controller,
+        description: EntityDescription,
+    ) -> None:
+        """Initialize."""
+        super().__init__(entry, coordinator, controller, description)
+
+        self._running: bool = False
+
+    @property
+    def status_key(self) -> str:
+        """Return the data key that contains the activity status."""
+        return "state"
+
+    def calculate_seconds_remaining(self) -> int:
+        """Calculate the number of seconds remaining."""
+        raise NotImplementedError
+
+    @callback
+    def update_from_latest_data(self) -> None:
+        """Update the state."""
+        data = self.coordinator.data[self.entity_description.uid]
+        now = utcnow()
+        run_state = RUN_STATE_MAP.get(data[self.status_key])
+
+        if run_state == RunStates.NOT_RUNNING:
+            if self._running:
+                # If we go from anything to not running, update the state to be right
+                # now (i.e., the time the zone stopped running):
+                self._attr_native_value = now
+                self._running = False
+            return
+
+        if run_state == RunStates.QUEUED:
+            # If the zone is queued, hold off on calcuating its completion time until
+            # it actually starts:
+            return
+
+        self._running = True
+
+        seconds_remaining = self.calculate_seconds_remaining()
+        new_timestamp = utcnow() + timedelta(seconds=seconds_remaining)
+
+        if self._attr_native_value:
+            assert isinstance(self._attr_native_value, datetime)
+            if (
+                new_timestamp - self._attr_native_value
+            ) < DEFAULT_ZONE_COMPLETION_TIME_WOBBLE_TOLERANCE:
+                # If the deviation between the previous and new timestamps is less than
+                # a "wobble tolerance," don't spam the state machine:
+                return
+
+        self._attr_native_value = new_timestamp
+
+
+class ProgramTimeRemainingSensor(TimeRemainingSensor):
+    """Define a sensor that shows the amount of time remaining for a program."""
+
+    def __init__(
+        self,
+        entry: ConfigEntry,
+        program_coordinator: DataUpdateCoordinator,
+        zone_coordinator: DataUpdateCoordinator,
+        controller: Controller,
+        description: EntityDescription,
+    ) -> None:
+        """Initialize."""
+        super().__init__(entry, program_coordinator, controller, description)
+
+        self._zone_coordinator = zone_coordinator
+
+    @property
+    def status_key(self) -> str:
+        """Return the data key that contains the activity status."""
+        return "status"
+
+    def calculate_seconds_remaining(self) -> int:
+        """Calculate the number of seconds remaining."""
+        program_data = self.coordinator.data[self.entity_description.uid]
```

Comment:
You could make `self.coordinator.data[self.entity_description.uid]` a property if you don't want to write it out every time.